### ISSUE TITLE
getOpenIntervals skips current week when used in the UTC timezone

### ISSUE THREAD

**Issue by spawn-guy** [opening-hours/opening_hours.js#171]
I have not yet started debugging the package' code, but

I have 2 Environments DEV and PRO. DEV is my local pc(GMT+2) , PRO is in amazon AWS eu-west-1 region (UTC).
i also have this schema "week 1-53/2 Tu 07:30-08:00".

```
let schema = 'week 1-53/2 Tu 07:30-08:00';
let fromMoment = moment();
let toMoment = moment().add(1, 'year');
let from = fromMoment.toDate();
let to = toMoment.toDate();

let OpeningHours = require('opening_hours');
let oh = new OpeningHours(schema, getnominatim());
let intervals = oh.getOpenIntervals(from, to);
```

so, the problem i am experiencing is that on DEV - _This_ week appears in "intervals" output. but on the PRO Server it is _NOT_ being calculated. the Intervals start _next_ week.

if on the PRO i set start time to "last sunday evening" (i.e. last week, -12hours as of now) - then this weeks calculates correctly.
it does NOT have effect on DEV env. this week is properly calculated.

this is server-only code. code is not mine. and i think i have custom "Nominatim" as i need to get results in NL locale.

for now i will be using "last sunday" as a start date, but i'd Looove to make a proper fix. Please tell me where to look for what.

**Comment by ypid** [comment]
Hm, I am not quite following. opening_hours.js does not deal with timezones. All times are local times. As for you bug with incorrect intervals I am not sure what the problem is there. Please let me know if you figure out what the problem is there and if there is a fix for this library. Thanks.

**Comment by spawn-guy** [comment]
no prob. people seem to not to understand questions from me in the first iteration. i'll explain with an example.

given schema: `week 1-53/2 Tu 07:30-08:00` = odd week Tuesday from 07:30 to 08:00.

object: `let from = moment();`
string: `Mon Jul 18 2016 14:33:31 GMT+0200 (W. Europe Daylight Time)` (log from DEV)

on DEV WIN PC produces intervals:
"_2016-07-19T07:30:00.000Z_", "2016-08-02T07:30:00.000Z", "2016-08-16T07:30:00.000Z"... etc.

on PRO Amazon Linux produces intervals:
"2016-08-02T07:30:00.000Z", "2016-08-16T07:30:00.000Z"... etc.

i.e. THIS week is Absent from the output (this is consistent.. as i have another rule for Wednesday Odd weeks)

however. if i use this:

object: `let from = moment().day(0).hour(23).minute(59).second(59);`
string: `Sun Jul 17 2016 23:59:59 GMT+0200 (W. Europe Daylight Time)`

then BOTH (all) environments produce correct, THIS week included, intervals: "_2016-07-19T07:30:00.000Z_", "2016-08-02T07:30:00.000Z", "2016-08-16T07:30:00.000Z"... etc.

for now i've used this as a "fix" we'll see how it goes

**Comment by ypid** [comment]
Ok. You said this code is intended to be executed on servers. Can you try to reproduce this in a web browser (→ Chromium or Firefox)? (I don’t quite trust Node.JS on MS Windows, never tried that myself for a reason :wink:)

**Comment by spawn-guy** [comment]
@ypid i am not able to verify it in the browsers (i can't also jsfiddle, due to httpS absence on oh.js CDN server).
however, i do get these missing weeks on the server side. pretty consistently.

Q: how do i make a proper test and run it on a headless linux?

**Comment by spawn-guy** [comment]
okay, @ypid ... some progress update :) the error is consistent. even on windows pc.
if the from is `new Date()` yes, the default "date-now" - then error is reproducible. 

```
test.addTest('Week range week without nominatim. Date(now)', [
    'week 01-53/2 Th 07:30-08:00',
], new Date('2016.11.09 16:59'), new Date('2016.11.10 16:59'), [
    ['2016.11.10 07:30', '2016.11.10 08:00'],
], 1800000, 0, false);

test.addTest('Week range week without nominatim. Date(now)', [
    'week 01-53/2 Th 07:30-08:00',
], new Date(), new Date('2016.11.10 16:59'), [
    ['2016.11.10 07:30', '2016.11.10 08:00'],
], 1800000, 0, false);
```

i've modified a few lines to accept Date objects.
```
this.runSingleTest = function (test_data_object) { /* {{{ */
        var name = test_data_object[0],
            value = test_data_object[1],
            first_value = test_data_object[2],
            from = test_data_object[3],
            to = test_data_object[4],
            expected_intervals = test_data_object[5],
            expected_durations = test_data_object[6],
            expected_weekstable = test_data_object[7],
            nominatimJSON = test_data_object[8],
            oh_mode = test_data_object[9];

        // fix from and to dates
        if (!(from instanceof Date)) {
            from = new Date(from);
        }
        if (!(to instanceof Date)) {
            to = new Date(to);
        }

        var ignored = typeof value !== 'string';
        if (ignored) {
            this.ignored.push(value);
            ignored = value[1];
            value = value[0];
        }

        var oh, intervals, durations, weekstable, prettified, intervals_ok, duration_ok, weekstable_ok, prettify_ok, crashed = true;

        var warnings;
        try {
            oh = new opening_hours(value, nominatimJSON, oh_mode);

            warnings = oh.getWarnings();

            intervals = oh.getOpenIntervals(from, to);
            durations = oh.getOpenDuration(from, to);
```

**Comment by spawn-guy** [comment]
okay, here is a funnier thing:

```
test.addTest('Week range week without nominatim. Date(now)', [
    'week 01-53/2 Th 07:30-08:00',
], Date(), new Date('2016.11.10 16:59'), [
    ['2016.11.10 07:30', '2016.11.10 08:00'],
], 1800000, 0, false);

test.addTest('Week range week without nominatim. Date(now)', [
    'week 01-53/2 Th 07:30-08:00',
], new Date(), new Date('2016.11.10 16:59'), [
    ['2016.11.10 07:30', '2016.11.10 08:00'],
], 1800000, 0, false);
```
*notice:* `Date()` vs `new Date()`
```
"Week range week without nominatim. Date(now)" for "week 01-53/2 Th 07:30-08:00": PASSED
"Week range week without nominatim. Date(now)" for "week 01-53/2 Th 07:30-08:00": FAILED, bad duration(s): 0,0, expected 1800000,0, bad intervals: 
(none)
expected:
[ '2016.11.10 07:30', '2016.11.10 08:00', false, undefined ],
1/2 tests passed. 1 did not pass.
```

**Comment by spawn-guy** [comment]
@ypid so what are the next steps?
shall i stop at this new testCase and make PR with just that case?
or shall i try to debug and fix the problem?

i.e. i have found a fix in for _my_ problem, but do we need to fix the library somehow?

**Comment by ypid** [comment]
@spawn-guy Thanks for looking into this! I am not quite following why you used `Date()` in the integration test examples you wrote? `this.runSingleTest` already creates a new `Date` object based on the string passed to it.

As I run the examples you gave 1 day after you posted them, I got different results.

```JavaScript
test.addTest('Week range week without nominatim. Date(now)', [
    'week 01-53/2 Th 07:30-08:00',
], '2016.11.10 05:00', '2016.11.10 16:59', [
    ['2016.11.10 07:30', '2016.11.10 08:00'],
], 1800000, 0, false);
```

Passes for me.

I probably don’t quite get it, sorry for that.

> shall i stop at this new testCase and make PR with just that case?
> or shall i try to debug and fix the problem?

Thanks for your help. I would prefer the "shall i try to debug and fix the problem?" :wink:

**Comment by spawn-guy** [comment]
@ypid Never fear - I is Here :)
I do understand the Frustration, but i pinpointed it and, actually, fixed it. and now i can explain and will do a PR in an hour, after prettifying. for now i didn't make a proper "dynamic" testCase, but i hope i will.

in short: in JS `new Date();` IS "Reactive" variable. `new Date('some value')` is a "Static" variable.
due to this - `getWeekNumber` - is, usually, mistaken.
it is mistaken 'cause `new Date()` Always holds Current time in **Milliseconds** since Epoch. so the value DOES change after a few "lines of code"(sic!). so the value of the Date is different in the beginning of the function and a few CPU-cycles later :|


with the following code i've been able to pinpoint and fix the issue.
```
function getWeekNumber(d) {
            // Copy date so don't modify original
            d = new Date(+d);
            d.setHours(0,0,0);
            // Set to nearest Thursday: current date + 4 - current day number
            // Make Sunday's day number 7
            d.setDate(d.getDate() + 4 - (d.getDay()||7));
            // Get first day of year
            var yearStart = new Date(d.getFullYear(),0,1);
            // Calculate full weeks to nearest Thursday
            var diff = (d - yearStart);
            var diffDiv = diff/86400000;
            var diffDivPlus = diffDiv+1;
            var diffDivPlusDiv = diffDivPlus/7;
            var diffDivPlusDivCeil =Math.ceil(diffDivPlusDiv);

            console.log("[d, diff ,diffDiv ,diffDivPlus ,diffDivPlusDiv,diffDivPlusDivCeil]",[d.getTime(), diff ,diffDiv ,diffDivPlus ,diffDivPlusDiv,diffDivPlusDivCeil]);

            return Math.ceil(( ( (d - yearStart) / 86400000) + 1)/7)
        }
```

it results in (ofcource the code is run multiple times, so i skipped a few iterations)
`new Date(Date())`: [d, diff ,diffDiv ,diffDivPlus ,diffDivPlusDiv,diffDivPlusDivCeil] [ 1478732400000, 27129600000, 314, 315, 45, 45 ]
`new Date()`: [d, diff ,diffDiv ,diffDivPlus ,diffDivPlusDiv,diffDivPlusDivCeil] [ 1478732400230,  27129600230,  314.00000266203705,  315.00000266203705,  45.000000380291006,  46 ]

so, `new Date(Date())` detects THIS week as **45**, but `new Date()` results in **46** due to fraction-of-a-second difference Ceiling results in +1.

the solution is to change the line `d.setHours(0,0,0);` to `d.setHours(0,0,0,0);` to make the millis also static :|

**Comment by ypid** [comment]
@spawn-guy Awesome work debugging this tricky issue! Thanks very much :+1:

Now that makes sense. I had similar problems in Python a few weeks ago. For Python there are static code analyzers that can find similar problems. This project is still lacking more advanced testing and checking like that. Not sure yet when I come around to improve the testing.

**Comment by spawn-guy** [comment]
as a side-note.. as we(this lib) are using moment.js... moment does have a .week() and .isoWeek() functions. might be wise to switch to using them,.. instead of ..

**Comment by ypid** [comment]
Hm, moment.js is currently not intended to be a hard dependency. The library started out without moment.js and in the light of #136 I would like to keep the hard dependencies to a minimum. But it is worth a thought to use methods already provided by other libs when available. Thanks!

**Comment by ypid** [comment]
Closed by: #191
